client/beefy/src/import.rs

```diff
@@ -0,0 +1,195 @@
+use std::{collections::HashMap, marker::PhantomData, sync::Arc};
+
+use sc_consensus::{
+	BlockCheckParams, BlockImport, BlockImportParams, ImportResult, JustificationImport,
+};
+
+use sc_client_api::backend::Backend;
+use sp_api::{ProvideRuntimeApi, TransactionFor};
+use sp_blockchain::well_known_cache_keys;
+use sp_consensus::Error as ConsensusError;
+use sp_runtime::{
+	generic::{BlockId, OpaqueDigestItemId},
+	traits::{Block as BlockT, Header as HeaderT, NumberFor},
+	Justification,
+};
+
+use beefy_primitives::{
+	crypto::{AuthorityId, Signature},
+	BeefyApi, ConsensusLog, ValidatorSet, BEEFY_ENGINE_ID,
+};
+
+use crate::{
+	justification::decode_and_verify_justification, notification::BeefyJustificationSender,
+	Client as BeefyClient,
+};
+
+/// Checks the given header for a consensus digest signalling a beefy authority set change
+/// and extracts it.
+pub fn find_beefy_authority_set_change<B: BlockT>(
+	header: &B::Header,
+) -> Option<ValidatorSet<AuthorityId>> {
+	let id = OpaqueDigestItemId::Consensus(&BEEFY_ENGINE_ID);
+
+	let filter_log = |log: ConsensusLog<AuthorityId>| match log {
+		ConsensusLog::AuthoritiesChange(change) => Some(change),
+		_ => None,
+	};
+
+	// find the first consensus digest with the right ID which converts to
+	// the right kind of consensus log.
+	header.digest().convert_first(|l| l.try_to(id).and_then(filter_log))
+}
+
+/// BeefyBlockImport
+/// Wraps a type `inner` that implements [`BlockImport`]
+pub struct BeefyBlockImport<Backend, Block: BlockT, Client, I> {
+	client: Arc<Client>,
+	inner: I,
+	justification_sender: BeefyJustificationSender<NumberFor<Block>, Signature>,
```

Comment:
Oh, and I'd forgot about this, sorry.

The code in the PR only validates BEEFY justifications in block imports and nothing else.

I am guessing there is a 2nd part or a future PR that uses this `justification_sender` to forward justifications from block import to the beefy gadget, right?

Reply:
Yeah, there should definitely be a second PR that makes use of this , but I don't understand fully how it would look like, was hoping to get insight on that after this was reviewed, but this just verifies beefy justfications if present and ensures there are justifications for authority set change blocks.